src/bases/createWidgetBase.ts

```diff
@@ -0,0 +1,201 @@
+import { ComposeFactory } from 'dojo-compose/compose';
+import createStateful from 'dojo-compose/bases/createStateful';
+import {
+	HNode,
+	DNode,
+	WNode,
+	Widget,
+	WidgetMixin,
+	WidgetState,
+	WidgetOptions
+} from 'dojo-interfaces/widgetBases';
+import { VNode, VNodeProperties } from 'dojo-interfaces/vdom';
+import { Factory } from 'dojo-interfaces/core';
+import { assign } from 'dojo-core/lang';
+import WeakMap from 'dojo-shim/WeakMap';
+import Map from 'dojo-shim/Map';
+import d from './../util/d';
+
+export interface WidgetFactory extends ComposeFactory<Widget<WidgetState>, WidgetOptions<WidgetState>> {}
+
+interface WidgetInternalState {
+	readonly id?: string;
+	dirty: boolean;
+	widgetClasses: string[];
+	cachedWidget?: VNode;
+	historicChildrenMap: Map<string | Factory<Widget<WidgetState>, WidgetOptions<WidgetState>>, Widget<WidgetState>>;
+	currentChildrenMap: Map<string | Factory<Widget<WidgetState>, WidgetOptions<WidgetState>>, Widget<WidgetState>>;
+};
+
+/**
+ * Internal state map for widget instances
+ */
+const widgetInternalStateMap = new WeakMap<Widget<WidgetState>, WidgetInternalState>();
+
+/**
+ * The counter for generating a unique ID
+ */
+let widgetCount = 0;
+
+function generateID(instance: Widget<WidgetState>): string {
+	const id = `widget-${++widgetCount}`;
+	instance.setState({ id });
+	return id;
+}
+
+function isWNode(child: DNode): child is WNode {
+	return child && (<WNode> child).factory !== undefined;
+}
+
+function realizeDNode(instance: Widget<WidgetState>, dNode: DNode) {
+	const internalState = widgetInternalStateMap.get(instance);
+	let child: HNode | Widget<WidgetState>;
+	if (isWNode(dNode)) {
+		const { factory, options: { id, state } } = dNode;
+		const childrenMapKey = id || factory;
+		const cachedChild = internalState.historicChildrenMap.get(childrenMapKey);
+		if (cachedChild) {
+			child = cachedChild;
+			if (state) {
+				child.setState(state);
+			}
+		} else {
+			child = factory(dNode.options);
+			child.own(child.on('invalidate', () => {
+				instance.invalidate();
+			}));
+			internalState.historicChildrenMap.set(childrenMapKey, child);
+			instance.own(child);
+		}
+		if (!id && internalState.currentChildrenMap.has(factory)) {
+			console.error('must provide unique keys when using the same widget factory multiple times');
+		}
+		internalState.currentChildrenMap.set(childrenMapKey, child);
+	}
+	else {
+		child = dNode;
+		if (child.children) {
+			child.children = child.children.map((child: DNode) => {
+				return realizeDNode(instance, child);
```

Comment:
could use implicit return here... `child.children.map((child) => realizeDNode(instance, child));`

Can children ever be not a `DNode`?  If this were broken out a separate function, it is likely you can tighten the types as well so that `child` could be inferred.